v1/ocagent.go

```
@@ -0,0 +1,345 @@
+	nodeInfo              *agentcommonpb.Node
+	grpcClientConn        *grpc.ClientConn
+	spanDataCh            chan *trace.SpanData
+	flushCh               chan bool
+	shutdownTraceExportCh chan bool
+}
+
+func NewExporter(opts ...ExporterOption) (*Exporter, error) {
+	exp, err := NewUnstartedExporter(opts...)
+	if err != nil {
+		return nil, err
+	}
+	if err := exp.Start(); err != nil {
+		return nil, err
+	}
+	return exp, nil
+}
+
+const spanDataBufferSize = 300
+
+func NewUnstartedExporter(opts ...ExporterOption) (*Exporter, error) {
+	e := new(Exporter)
+	for _, opt := range opts {
+		opt.withExporter(e)
+	}
+	if e.agentPort <= 0 {
+		e.agentPort = DefaultAgentPort
+	}
+	if e.agentHost == "" {
+		e.agentHost = DefaultAgentHost
+	}
+	e.flushCh = make(chan bool)
+	e.shutdownTraceExportCh = make(chan bool, 1)
+	e.spanDataCh = make(chan *trace.SpanData, spanDataBufferSize)
+	e.nodeInfo = createNodeInfo(e.serviceName)
+	return e, nil
+}
+
+const (
+	maxInitialConfigRetries = 10
+	maxInitialTracesRetries = 10
+)
+
+// Start dials to the agent, establishing a connection to it. It also
+// initiates the Config and Trace services by sending over the initial
+// messages that consist of the node identifier. Start performs a best case
+// attempt to try to send the initial messages, by applying exponential
+// backoff at most 10 times.
+func (ae *Exporter) Start() error {
+	ae.mu.Lock()
+	defer ae.mu.Unlock()
+
+	err := ae.doStartLocked()
+	if err == nil {
+		return nil
+	}
+
+	// Otherwise we have an error and should clean up to avoid leaking resources.
+	if ae.grpcClientConn != nil {
+		ae.grpcClientConn.Close()
+	}
+
+	return err
+}
+
+func (ae *Exporter) doStartLocked() error {
+	if ae.started {
+		return nil
+	}
+
+	// Now start it
+	addr := fmt.Sprintf("%s:%d", ae.agentHost, ae.agentPort)
+	var dialOpts []grpc.DialOption
+	if ae.canDialInsecure {
+		dialOpts = append(dialOpts, grpc.WithInsecure())
+	}
+	cc, err := grpc.Dial(addr, dialOpts...)
```

I think if the agent is not started this will fail. Am I wrong? If that is the case the retry logic should be here.

---

Now I recall why I didn't choose to add connection retry logic here. Using exponential backoff, this can possibly extend application start time by infinitely long durations. Perhaps I'll provide an option to fail fast.

---

>I think if the agent is not started this will fail. Am I wrong? If that is the case the retry logic should be here.

Not really, by default grpc.Dial returns without checking if the connection succeeded or not https://github.com/grpc/grpc-go/blob/46ee6abebf62241b49973422c42250327bb94169/clientconn.go#L105-L121

in the case of a bad connection, with the usual grpc.Dial, it'll return and then fail when setting up the client services. There is a grpc.DialOption called `WithDial` but it seems to block even on bad connections and doesn't fail fast -- so we don't want that, I think the retry logic should be with creating the various grpc client services.